lib/types/iprange.go

```diff
@@ -0,0 +1,262 @@
+package types
+
+import (
+	"encoding/binary"
+	"fmt"
+	"hash/fnv"
+	"net"
+	"strconv"
+	"strings"
+)
+
+// selectMode tells how IP is selected in a IPBlock splited from --ip option
+
+type selectMode uint
+
+const (
+	roundRobin selectMode = iota
+	random
+)
+
+// IPBlock represents a continuous segment of IP addresses
+// ip     - the first IP in this address segment. CIDR can skip lower IPs
+// weight - how many IPs will be chosen from this segment. default is all
+// mode   - method to choose IP, sequential(0) or random(1). default is 0
+// hostStart - uint64 conversion of fistIP[8:16]
+// netStart  - uint64 conversion of fistIP[0:8]
+// hostN  - number of IPv4 addrs (IPv6 hosts) in this address segment
+// netN   - number of IPv6 nets in this address segment, always 1 if IPv4
+type IPBlock struct {
+	ip        net.IP
+	hostStart uint64
+	netStart  uint64
+	hostN     uint64
+	netN      uint64
+	weight    uint64
+	split     uint64
+	mode      selectMode
+}
+
+// IPPool represent a slice of IPBlocks
+type IPPool []IPBlock
+
+// GetIPBlock return a struct of a sequential IP range
+// support single IP, range format by '-' or CIDR by '/'
+// support up to 2^32 IPv4 addresses and 2^64 * 2^64 IPv6 addesses
+// for IPv6, high 64 bits is independent to low 64 bits if input > 64 bits
+// this is useful if you want to test some IPv6 nets without traveling 2^64 hosts
+func GetIPBlock(s string) *IPBlock {
+	for i := 0; i < len(s); i++ {
+		switch s[i] {
+		case '-':
+			return ipBlockFromRange(s)
+		case '/':
+			return ipBlockFromCIDR(s)
+		}
+	}
+	return ipBlockFromRange(s + "-" + s)
+}
+
+func hashIDToUint64(id uint64) uint64 {
+	a, b := fnv.New64(), make([]byte, 8)
+	binary.LittleEndian.PutUint64(b, id)
+	if _, err := a.Write(b); err != nil {
+		return 0
+	}
+	return a.Sum64()
+}
+
+func ipUint64(ip net.IP) (uint64, uint64) {
+	ip128 := ip.To16()
+	net64 := binary.BigEndian.Uint64(ip128[:8])
+	host64 := binary.BigEndian.Uint64(ip128[8:])
+	return net64, host64
+}
+
+func ipBlockFromRange(s string) *IPBlock {
+	ss := strings.SplitN(s, "-", 2)
+	ip0Str, ip1Str := strings.TrimSpace(ss[0]), strings.TrimSpace(ss[1])
+	ip0, ip1 := net.ParseIP(ip0Str), net.ParseIP(ip1Str) // len(ParseIP())==16
+	if ip0 == nil || ip1 == nil {
+		fmt.Println("Wrong IP range format: ", s)
+		return nil
+	}
+	n0, h0 := ipUint64(ip0)
+	n1, h1 := ipUint64(ip1)
+	if (n0 > n1) || (h0 > h1) {
```

Comment:
This is by design for IPv6 range tests. I divide 128 bit to two 64bit, host 64bit and network 64bit, each 64bit will change range independently. This is to test some IPv6 network range (if not, k6 must travel through all 2^64 host IP addresses, that is terrible) 

for IPv4, only 32bit hides in host 64bit, so it will not impact anything.